### ISSUE TITLE
python.app/Contents/MacOS/python is a copy, not a sym-link

### ISSUE THREAD

**Issue by pauldmccarthy** [conda-forge/python.app-feedstock#25]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Howdy, I am chasing down a shared library issue, and have discovered that when `python.app` is installed into an environment, `$PREFIX/python.app/Contents/MacOS/python` ends up as a copy of `$PREFIX/bin/python`, rather than a sym-link to it.

I think this is because of the logic in `post-link.sh` - the python.app framework directory is initially installed to `$PREFIX/pythonapp`, but then in `post-link.sh` is copied to `$PREFIX/python.app` using `cp -r`.

On macOS, `cp -r` follows, rather than copies, sym-links. You can test this out yourself:

```
$ mkdir test && cd test
$ mkdir bin 
$ echo "hello" > bin/python
$ mkdir -p pythonapp/Contents/MacOS/
$ cd pythonapp/Contents/MacOS/
$ ln -s ../../../bin/python  python
$ cd ../../../
$ ls -l pythonapp/Contents/MacOS/

lrwxr-xr-x  1 paulmc  staff  19 10 Jan 12:24 python@ -> ../../../bin/python

$ cp -r pythonapp python.app
$ ls -l python.app/Contents/MacOS/python

-rw-r--r--  1 paulmc  staff  6 10 Jan 12:24 python.app/Contents/MacOS/python
```

The macOS manpage for `cp` discourages the use of `-r` for this reason, and instead
provides the `-R` option:

```
$ cp -R pythonapp python.appR
$ ls -l python.appR/Contents/MacOS/python

lrw-r--r--  1 paulmc  staff  19 10 Jan 12:25 python.appR/Contents/MacOS/python@ -> ../../../bin/python
```

So this issue can be resolved by changing `post-link.sh` to use `cp -R` instead of `cp -r.`

I'm happy to follow up with a PR if you agree. However, I'm also curious as to why the `post-link.sh` is even necessary - why does `build.sh` create `pythonapp` and not `python.app`?

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
 # packages in environment at /Users/paulmc/test.env:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                h0d85af4_4    conda-forge
ca-certificates           2021.10.8            h033912b_0    conda-forge
libffi                    3.4.2                h0d85af4_5    conda-forge
libzlib                   1.2.11            h9173be1_1013    conda-forge
ncurses                   6.2                  h2e338ed_4    conda-forge
openssl                   3.0.0                h0d85af4_2    conda-forge
pip                       21.3.1             pyhd8ed1ab_0    conda-forge
python                    3.10.1          h38b4d05_2_cpython    conda-forge
python.app                1.3             py310he24745e_6    conda-forge
python_abi                3.10                    2_cp310    conda-forge
readline                  8.1                  h05e3726_0    conda-forge
setuptools                60.5.0                   pypi_0    pypi
sqlite                    3.37.0               h23a322b_0    conda-forge
tk                        8.6.11               h5dbffcc_1    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                haf1e3a3_1    conda-forge
zlib                      1.2.11            h9173be1_1013    conda-forge
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : None
       user config file : /Users/paulmc/.condarc
 populated config files : 
          conda version : 4.10.3
    conda-build version : not installed
         python version : 3.9.5.final.0
       virtual packages : __osx=10.15.7=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /Users/paulmc/miniconda3  (writable)
      conda av data dir : /Users/paulmc/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/paulmc/miniconda3/pkgs
                          /Users/paulmc/.conda/pkgs
       envs directories : /Users/paulmc/miniconda3/envs
                          /Users/paulmc/.conda/envs
               platform : osx-64
             user-agent : conda/4.10.3 requests/2.25.1 CPython/3.9.5 Darwin/19.6.0 OSX/10.15.7
                UID:GID : 503:20
             netrc file : None
           offline mode : False
```
</details>

**Comment by pauldmccarthy** [comment]
As an aside, my shared library issue is that `$PREFIX/python.app/Contents/MacOS/python` (when installed as a copy of `$PREFIX/bin/python`) doesn't seem to be able to find shared libraries installed into `$PREFIX/lib/`. 

For example. [`rtree`](https://github.com/conda-forge/rtree-feedstock) is a Python wrapper around the [`libspatialindex`](https://github.com/conda-forge/libspatialindex-feedstock) C library, which is installed to `$PREFIX/lib/libspatialindex_c.dylib`:

```
$ conda create -c conda-forge -p rtree.env -y python python.app rtree
```

<details>

```
Collecting package metadata (current_repodata.json): done                                                                                                                                      
Solving environment: done                                                                                                                                                                      
                                                                                                                                                                                               
                                                                                                                                                                                               
==> WARNING: A newer version of conda exists. <==                                                                                                                                              
  current version: 4.10.3                                                                                                                                                                      
  latest version: 4.11.0                                                                                                                                                                       
                                                                                                                                                                                               
Please update conda by running                                                                                                                                                                 
                                                                                                                                                                                               
    $ conda update -n base -c defaults conda                                                                                                                                                   
                                                                                                                                                                                               
                                                                                                                                                                                               
                                                                                                                                                                                               
## Package Plan ##                                                                                                                                                                             
                                                                                                                                                                                               
  environment location: /Users/paulmc/rtree.env                                                                                                                                                
                                                                                                                                                                                               
  added / updated specs:                                                                                                                                                                       
    - python                                                                                                                                                                                   
    - python.app                                                                                                                                                                               
    - rtree                                                                                                                                                                                    
                                                                                                                                                                                               
                                                                                                                                                                                               
The following packages will be downloaded:                                                                                                                                                     
                                                                                                                                                                                               
    package                    |            build
    ---------------------------|-----------------
    rtree-0.9.7                |  py310had9ce37_3          46 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          46 KB

The following NEW packages will be INSTALLED:

  bzip2              conda-forge/osx-64::bzip2-1.0.8-h0d85af4_4
  ca-certificates    conda-forge/osx-64::ca-certificates-2021.10.8-h033912b_0
  libcxx             conda-forge/osx-64::libcxx-12.0.1-habf9029_1
  libffi             conda-forge/osx-64::libffi-3.4.2-h0d85af4_5
  libspatialindex    conda-forge/osx-64::libspatialindex-1.9.3-he49afe7_4
  libzlib            conda-forge/osx-64::libzlib-1.2.11-h9173be1_1013
  ncurses            conda-forge/osx-64::ncurses-6.2-h2e338ed_4
  openssl            conda-forge/osx-64::openssl-3.0.0-h0d85af4_2
  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0
  python             conda-forge/osx-64::python-3.10.1-h38b4d05_2_cpython
  python.app         conda-forge/osx-64::python.app-1.3-py310he24745e_6
  python_abi         conda-forge/osx-64::python_abi-3.10-2_cp310
  readline           conda-forge/osx-64::readline-8.1-h05e3726_0
  rtree              conda-forge/osx-64::rtree-0.9.7-py310had9ce37_3
  setuptools         conda-forge/osx-64::setuptools-60.5.0-py310h2ec42d9_0
  sqlite             conda-forge/osx-64::sqlite-3.37.0-h23a322b_0
  tk                 conda-forge/osx-64::tk-8.6.11-h5dbffcc_1
  tzdata             conda-forge/noarch::tzdata-2021e-he74cb21_0
  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0
  xz                 conda-forge/osx-64::xz-5.2.5-haf1e3a3_1
  zlib               conda-forge/osx-64::zlib-1.2.11-h9173be1_1013



Downloading and Extracting Packages
rtree-0.9.7          | 46 KB     | #################################################################################################################################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate /Users/paulmc/rtree.env
#
# To deactivate an active environment, use
#
#     $ conda deactivate
```

</details>

Running `import rtree` will result in a call to `ctypes.utils.find_library('spatialindex_c')` (where `libspatialindex_c.dylib` is installed into `$PREFIX/lib/`). When calling the `python.app` interpreter directly, the library cannot be found:

```
$ ./rtree.env/python.app/Contents/MacOS/python -c "import rtree"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/paulmc/rtree.env/python.app/Contents/lib/python3.10/site-packages/rtree/__init__.py", line 9, in <module>
    from .index import Rtree, Index  # noqa
  File "/Users/paulmc/rtree.env/python.app/Contents/lib/python3.10/site-packages/rtree/index.py", line 6, in <module>
    from . import core
  File "/Users/paulmc/rtree.env/python.app/Contents/lib/python3.10/site-packages/rtree/core.py", line 77, in <module>
    rt.Error_GetLastErrorNum.restype = ctypes.c_int
  File "/Users/paulmc/rtree.env/python.app/Contents/lib/python3.10/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/Users/paulmc/rtree.env/python.app/Contents/lib/python3.10/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(RTLD_DEFAULT, Error_GetLastErrorNum): symbol not found
```

Calling `$PREFIX/bin/python.app` or `$PREFIX/bin/pythonw` works, however, as does replacing `$PREFIX/python.app/Contents/MacOS/python` with a sym-link to `$PREFIX/bin/python`.

**Comment by hoechenberger** [comment]
Hello @pauldmccarthy, this sounds reasonable to me, would you like to make a PR?

**Comment by ChrisBarker-NOAA** [comment]
Do make sure you test carefully.

IIUC, the entire point of Python.app is that Apple requires the executable to be inside an app bundle. A symlink may not satisfy that requirement.

**Comment by pauldmccarthy** [comment]
Hi @ChrisBarker-NOAA, thanks for chiming in. I've submitted a PR for this (#26), if you'd like to take a look.

The symlink issue was originally discussed at #8 and #25, where it was deemed that sym-linking appears to work ([e.g.](https://github.com/conda-forge/python.app-feedstock/issues/8#issuecomment-596152492)). But I'm guessing that nobody actually checked that the sym-link was being installed/created correctly.

I've tested a [wxPython application](https://github.com/conda-forge/fsleyes-feedstock) (for which I am the developer/maintainer) on macOS Catalina, and the sym-link approach works fine, when calling all of `$PREFIX/bin/pythonw`, `$PREFIX/bin/python.app`, and `$PREFIX/python.app/Contents/MacOS/python`. 

I'm happy to do any more checks you can think of, although I only have access to the one macOS machine. 

I'd be happy to knock up a simple unit test to check GUI functionality although I'm unsure if the conda-forge macOS build infrastructure has any support for GUIs. I suspect not, as I remember having trouble getting unit / integration tests working for my application. 

Furthermore, it seems that `Tkinter` works just fine with `$PREFIX/bin/python` (i.e. `python.app` does not seem to be required to display `Tkinter` GUIs). So any test would need to use e.g. wxPython (which I'm familiar with) or, perhaps better, Cocoa via pyobjc (which I'm not at all familiar with).

**Comment by pauldmccarthy** [comment]
As far as I can tell, the [existing unit test](https://github.com/conda-forge/python.app-feedstock/blob/dbeafd9c6f9de264f85ddd8c671e4ff3c6d2ef88/recipe/run_test.sh) is always going to pass. What would really need to be tested would be to ensure that the python versions remain consistent across an upgrade. I doubt that this would be possible within the testing infrastructure available with `conda build`, but I might be wrontg there?

**Comment by ChrisBarker-NOAA** [comment]
@pauldmccarthy: if it works, it works, thanks!

This is SO confusing!

And i'm still confused as to how tkInter works without the app bundle -- it sure would be nice to apply that same trick to wxPython and PyQT (and pyObjC?).

Actually, I think that unit test did fail at some point, as the copy of the executable was being installed by hand, and someone might have forgotten to update it when the bundle was updated. But yes, it probably can't fail with the current build script.

Anyway, thanks for finding that old issue -- turns out that I raised, and resolved, the same issue two years ago. Oh well, carry on.
